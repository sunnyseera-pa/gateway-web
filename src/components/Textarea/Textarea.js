/** @jsx jsx */
import { cx } from '@emotion/css';
import { css, jsx } from '@emotion/react';
import { Form, InputGroup } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { addCommonPropTypes } from '../../configs/propTypes';
import useCommonStyles from '../../hooks/useCommonStyles';
import * as styles from './Textarea.styles';
import TextareaAutosize from 'react-textarea-autosize';

const Textarea = ({
	className,
	autosize,
	label,
	charCountLength,
	variant,
	value,
	mt,
	mb,
	ml,
	mr,
	width,
	minWidth,
	maxWidth,
	inputRef,
	id,
	...outerProps
}) => {
	const commonStyles = useCommonStyles({ mt, mb, ml, mr, width, width, minWidth, maxWidth });

	return (
		<Form.Group controlId={id} css={styles.formGroup}>
			<Form.Label css={styles.label}>
				{label}
				{!!charCountLength && (
					<span css={styles.charCount}>
						{' '}
						({value ? value.length : 0}/{charCountLength})
					</span>
				)}
			</Form.Label>

			<InputGroup css={styles.inputGroup({ variant })} className={cx('ui-Textarea', className, commonStyles)}>
				{autosize && <TextareaAutosize type='text' value={value} {...outerProps} ref={inputRef} />}
				{!autosize && <Form.Control as='textarea' {...outerProps} ref={inputRef} />}
			</InputGroup>
		</Form.Group>
	);
};

Textarea.defaultProps = {
	autosize: false,
	variant: 'primary',
};

Textarea.propTypes = addCommonPropTypes({
	label: PropTypes.node,
	placeholder: PropTypes.string,
	value: PropTypes.string,
	maxlength: PropTypes.number,
	charCountLength: PropTypes.number,
	onChange: PropTypes.func,
	inputRef: PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
	variant: PropTypes.oneOf(['primary', 'secondary']),
	id: PropTypes.string,
	name: PropTypes.string,
	autosize: PropTypes.bool,
});

export default Textarea;
