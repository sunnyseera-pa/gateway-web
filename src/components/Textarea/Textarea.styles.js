import { css } from '@emotion/react';
import { mixins } from '../Input/Input.styles';

export const inputGroup = ({ variant }) => theme => {
	const {
		colors,
		components: {
			Textarea: { variants },
		},
	} = theme;

	return css`
		width: 100%;

		textarea {
			${mixins.input({ variant })({ colors, variants })}
			width: 100%;
		}
	`;
};

export const formGroup = css`
	display: flex;
	flex-direction: column;
`;

export const label = css`
	display: flex;
`;

export const charCount = ({ colors: { grey700Alt } }) => css`
	flex-grow: 1;
	text-align: right;
	color: ${grey700Alt};
`;
