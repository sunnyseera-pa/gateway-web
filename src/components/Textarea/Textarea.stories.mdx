import { Meta, Story, Canvas } from '@storybook/addon-docs';

import Textarea from './Textarea';

<Meta title='Core/Textarea' component={Textarea} />

<Canvas>
	<Story name='Basic'>{props => <Textarea label='Label' />}</Story>
	<Story name='With count'>{props => <Textarea label='Label' charCountLength={255} />}</Story>
	<Story name='Autosize'>{props => <Textarea label='Label' autosize />}</Story>
</Canvas>
