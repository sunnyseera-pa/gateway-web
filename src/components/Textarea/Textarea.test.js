import { render, waitFor } from '@testing-library/react';
import React from 'react';
import Textarea from '.';

const mockResize = jest.fn();

jest.mock('react-textarea-autosize', () => props => {
	mockResize(props);
	return null;
});

const props = {
	label: 'Sample label',
	value: 'Lorem ipsum',
	charCountLength: 255,
	onChange: jest.fn(),
};

let wrapper;

describe('Given the Textarea component', () => {
	describe('When it is rendered', () => {
		beforeAll(() => {
			wrapper = render(<Textarea {...props} />, {
				wrapper: Providers,
			});
		});

		it('Then matches the previous snapshot', () => {
			expect(wrapper.container).toMatchSnapshot();
		});

		it('Then should use a standard textarea', () => {
			expect(mockResize).not.toHaveBeenCalled();
		});

		it('Then has a label with a count', () => {
			expect(wrapper.container.querySelector('.form-label').textContent).toEqual('Sample label (11/255)');
		});

		describe('And it is set to not show char count', () => {
			it('Then autosize with the correct arguments', () => {
				wrapper.rerender(<Textarea {...props} charCountLength={0} />, {
					wrapper: Providers,
				});

				expect(wrapper.container.querySelector('.form-label').textContent).toEqual('Sample label');
			});
		});

		describe('And it is set to autosize', () => {
			it('Then calls autosize with the correct arguments', () => {
				wrapper.rerender(<Textarea {...props} autosize />, {
					wrapper: Providers,
				});

				expect(mockResize.mock.calls[0][0]).toMatchObject({ type: 'text', value: 'Lorem ipsum', onChange: expect.any(Function) });
			});
		});
	});
});
