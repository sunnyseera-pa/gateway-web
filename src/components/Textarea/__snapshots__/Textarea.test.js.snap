// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Textarea component When it is rendered Then matches the previous snapshot 1`] = `
<div>
  <div
    class="css-hdsa3t-Textarea form-group"
  >
    <label
      class="css-1rtys8a-Textarea form-label"
    >
      Sample label
      <span
        class="css-1xasuay-Textarea"
      >
         
        (
        11
        /
        255
        )
      </span>
    </label>
    <div
      class="ui-Textarea css-1sivg29 css-1klfiiz-Textarea input-group"
    >
      <textarea
        class="form-control"
      />
    </div>
  </div>
</div>
`;
